### PR TITLE
feature(coq): support :standard in coq.pp (modules) field

### LINKED ISSUE #2414: coq.pp :standard?
Would it be possible for `coq.pp` to support `:standard` and globs? (with `:standard` meaning `*.mlg`)

### PR DESCRIPTION
**Pull request by Alizter**
fixes #2414 

- [x] Depends on #6228
- [x] Changelog
- [x] Doc

### CODE REVIEW

File: test/blackbox-tests/test-cases/coq/coqpp.t/run.t [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  $ ls
+  dune
+  dune-project
+  gram.mlg
+  marg.mlg
+
+  $ dune build
+  File "gram.ml", line 2, characters 8-30:
+  2 | let _ = Mltop.add_known_module __coq_plugin_name
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/coq/coqpp.t/run.t:18]: @ejgallego Do you know what we can do about this?

Comment (ejgallego) [reply test/blackbox-tests/test-cases/coq/coqpp.t/run.t:18]: You need to depend on `coq-core.vernac` on the library stanza, no other way around that as `coqpp` will generate Coq-specific code.

Comment (Alizter) [reply test/blackbox-tests/test-cases/coq/coqpp.t/run.t]: Thanks

**Review by christinerose — APPROVED** [review]

**Review by christinerose — APPROVED** [review]

File: doc/coq.rst [hunk 2]
```diff
@@ -304,21 +304,15 @@ coq.pp
...
 Authors of Coq plugins often need to write ``.mlg`` files to extend the Coq
 grammar. Such files are preprocessed with the ``coqpp`` binary. To help plugin
-authors avoid writing boilerplate, we provide a ``(coqpp ...)`` stanza:
+authors avoid writing boilerplate, we provide a ``(coq.pp ...)`` stanza:
 
 .. code:: scheme
 
-    (coq.pp (modules <mlg_list>))
+    (coq.pp
+     (modules <ordered_set_lang>))
```
Comment (ejgallego) [inline doc/coq.rst:312]: If I understand the code correctly, this should be a glob.

File: src/dune_rules/coq_stanza.mli [hunk 3]
```diff
@@ -41,7 +41,7 @@ end
 
 module Coqpp : sig
   type t =
-    { modules : string list
+    { modules : Predicate_lang.Glob.t
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply src/dune_rules/coq_stanza.mli:44]: Mmmm, why do we need them? Can't we just do like we do for `modules` in `(coq.theory ...)` ?

[the remark this answers is not in the document]
Comment (ejgallego) [reply src/dune_rules/coq_stanza.mli:44]: I'm not against using a glob here (tho I'm not sure I understand the implications), but IMHO we should keep the `modules` field consistent among stanzas.

File: doc/coq.rst (hunk 2 above)
Comment (Alizter) [reply doc/coq.rst:312]: It's easier to fetch the mlg sources using the predicate lang than it is with the ordered set one, so I would rather keep it like this.

File: src/dune_rules/coq_stanza.mli (hunk 3 above)
[the remark this answers is not in the document]
Comment (Alizter) [reply src/dune_rules/coq_stanza.mli:44]: It's easier to fetch the mlg sources using the predicate lang than it is with the ordered set one, so I would rather keep it like this.

[the remark this answers is not in the document]
Comment (ejgallego) [reply src/dune_rules/coq_stanza.mli:44]: Mmm, may be easier, but I'm still not very convinced that globs are what we need to have here; I'd rather prime correctness.

More generally, the whole idea of the `Coq_sources` modules is to have fetchers go thru the `Coq_sources.t` type, which in this case we avoid, so not sure this is indeed the right approach.

@rgrinberg what do you think?

[the remark this answers is not in the document]
Comment (Alizter) [reply src/dune_rules/coq_stanza.mli]: OK I've changed it to Ordered_set_lang.

File: doc/coq.rst (hunk 2 above)
Comment (Alizter) [reply doc/coq.rst:312]: It is now ordered set lang behind the scenes.

File: src/dune_rules/coq_stanza.mli (hunk 3 above)
[the remark this answers is not in the document]
Comment (ejgallego) [reply src/dune_rules/coq_stanza.mli]: Note that I'm not sure which is the right one, I don't have enough expertise on this part of Dune.

Also, still the question about whether that function should return a `Coq_sources.t` is IMHO valid, but I think it was Rudi who introduced this module, so he may know (I don't).

[the remark this answers is not in the document]
Comment (ejgallego) [reply src/dune_rules/coq_stanza.mli]: The PR looks good to me but indeed I feel like I don't have enough knolegde to review it in full.

[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_rules/coq_stanza.mli]: > More generally, the whole idea of the Coq_sources modules is to have fetchers go thru the Coq_sources.t type, which in this case we avoid, so not sure this is indeed the right approach.

The PR is implemented correctly. Or at least it is consistent with how we handle things with OCaml. `Coq_sources.t` is constructed after `(include_subdirs qualified)` has been computed, so it's correct to evaluate the .mlg rules before constructing the `Coq_sources.t`.

I have no opinion on whether the feature is a good idea, but we do have similar features on the OCaml side.

**Review by ejgallego — APPROVED** [review]

File: src/dune_rules/coq_stanza.mli (hunk 3 above)
[the remark this answers is not in the document]
Comment (ejgallego) [reply src/dune_rules/coq_stanza.mli]: Cool, thanks for the help @rgrinberg

**Comment by ejgallego** [comment]
Contrary to what the PR description says, this PR doesn't guard the change on the 0.7 Coq build language; but IMHO that's not a big deal.

**Comment by Alizter** [comment]
@ejgallego I don't think it would be easy to guard the `:standard` clause and probably not worth it since mlg use is so little. The PR description was really referencing the fact that when I was bumping to 0.7 I had a merge conflict so this needed to wait.

**Pull request merged**